csharp/src/Apache.Arrow/Ipc/ArrowReaderImplementation.cs

```diff
@@ -0,0 +1,207 @@
+
+        protected static T ReadMessage<T>(ByteBuffer bb)
+            where T : struct, IFlatbufferObject
+        {
+            var returnType = typeof(T);
+            var msg = Flatbuf.Message.GetRootAsMessage(bb);
+
+            if (MatchEnum(msg.HeaderType, returnType))
+            {
+                return msg.Header<T>().Value;
+            }
+            else
+            {
+                throw new Exception($"Requested type '{returnType.Name}' " +
+                                    $"did not match type found at offset => '{msg.HeaderType}'");
+            }
+        }
+
+        private static bool MatchEnum(Flatbuf.MessageHeader messageHeader, Type flatBuffType)
+        {
+            switch (messageHeader)
+            {
+                case Flatbuf.MessageHeader.RecordBatch:
+                    return flatBuffType == typeof(Flatbuf.RecordBatch);
+                case Flatbuf.MessageHeader.DictionaryBatch:
+                    return flatBuffType == typeof(Flatbuf.DictionaryBatch);
+                case Flatbuf.MessageHeader.Schema:
+                    return flatBuffType == typeof(Flatbuf.Schema);
+                case Flatbuf.MessageHeader.Tensor:
+                    return flatBuffType == typeof(Flatbuf.Tensor);
+                case Flatbuf.MessageHeader.NONE:
+                    throw new ArgumentException("MessageHeader NONE has no matching flatbuf types", nameof(messageHeader));
+                default:
+                    throw new ArgumentException($"Unexpected MessageHeader value", nameof(messageHeader));
+            }
+        }
+
+        protected RecordBatch CreateArrowObjectFromMessage(Flatbuf.Message message, ByteBuffer bodyByteBuffer)
+        {
+            switch (message.HeaderType)
+            {
+                case Flatbuf.MessageHeader.Schema:
+                    // TODO: Read schema and verify equality?
+                    break;
+                case Flatbuf.MessageHeader.DictionaryBatch:
+                    // TODO: not supported currently
+                    Debug.WriteLine("Dictionaries are not yet supported.");
+                    break;
+                case Flatbuf.MessageHeader.RecordBatch:
+                    var rb = message.Header<Flatbuf.RecordBatch>().Value;
+                    var arrays = BuildArrays(Schema, bodyByteBuffer, rb);
+                    return new RecordBatch(Schema, arrays, (int)rb.Length);
+                default:
+                    // NOTE: Skip unsupported message type
+                    Debug.WriteLine($"Skipping unsupported message type '{message.HeaderType}'");
+                    break;
+            }
+
+            return null;
+        }
+
+        private IEnumerable<IArrowArray> BuildArrays(
+            Schema schema,
+            ByteBuffer messageBuffer,
+            Flatbuf.RecordBatch recordBatchMessage)
+        {
+            var arrays = new List<ArrayData>();
```

Comment:
This seems to have some unnecessary costs:
- We're building up a List but we appear to know the exact length needed in advance.
- We're building up one thing but then using LINQ to convert it to something else.

Instead, why not do something like:
```C#
var arrays = new IArrowArray[recordBachMessage.NodesLength];
int bufferIndex = 0;
for (int n = 0; n < recordBackMessage.NodesLength; n++)
{
    int field = schema.GetFieldByIndex(n);
    var fieldNode = recordBatchMessage.Nodes(n).GetValueOrDefault();
    arrays[n] = field.DataType.IsFixedPrimitive() ?
        LoadPrimitiveField(field, fieldNode, recordBatchMessage, messageBuffer, ref bufferIndex) :
        LoadVariableField(field, fieldNode, recordBatchMessage, messageBuffer, ref bufferIndex);
}
return arrays;
```

Reply:
Will fix.

Reply:
I've re-written this section to not allocate so much. And it shows on my benchmarks:

|                      Method |       Allocated Memory/Op |
|----------------------------:|--------------------:|
|       Before |   65.14 KB |
|       After |    57.33 KB |